File: tiq-e/dynamic-programming/4/main.ts
import {performance} from "perf_hooks";

function rob(nums: number[]): number {
    let dp: number[] = [0, 0]
    for (let i = 0; i < nums.length; i++) {
        [dp[0], dp[1]] = [dp[1], Math.max(dp[1], dp[0] + nums[i])]
    }
    return dp[1]
};

const n = [2, 7, 9, 3, 1];

const start = performance.now();
const result = rob(n);
const end = performance.now();
console.log("Duration:", end - start);
console.log("Result:", result);

File: tiq-e/dynamic-programming/4/main.go
package main

import (
	"fmt"
	"time"
)

func rob(nums []int) int {
	var prev1, prev2 int
	for _, num := range nums {
		prev1, prev2 = max(prev2+num, prev1), prev1
	}

	return prev1
}

func max(n1, n2 int) int {
	if n1 > n2 {
		return n1
	}
	return n2
}

func main() {
	var n = []int{2, 7, 9, 3, 1}

	var start = time.Now()
	var result = rob(n)
	var end = time.Since(start)
	fmt.Println("Duration: ", end)
	fmt.Println("Result: ", result)
}
